native/sim_unicorn.cpp

```diff
@@ -2235,82 +2426,27 @@ static void hook_block(uc_engine *uc, uint64_t address, int32_t size, void *user
 	state->interrupt_handled = false;
+	auto curr_simos = state->get_simos();
+	auto angr_mode = state->get_angr_mode();
 
-	if (state->arch == UC_ARCH_X86 && intno == 0x80) {
-		// this is the ultimate hack for cgc -- it must be enabled by explitly setting the transmit sysno from python
-		// basically an implementation of the cgc transmit syscall
-
-		for (auto sr : state->symbolic_registers) {
-			// eax,ecx,edx,ebx,esi
-			if ((sr >= 8 && sr <= 23) || (sr >= 32 && sr <= 35)) return;
-		}
-
-		uint32_t sysno;
-		uc_reg_read(uc, UC_X86_REG_EAX, &sysno);
-		//printf("SYSCALL: %d\n", sysno);
-		if (sysno == state->transmit_sysno) {
-			//printf(".. TRANSMIT!\n");
-			uint32_t fd, buf, count, tx_bytes;
-
-			uc_reg_read(uc, UC_X86_REG_EBX, &fd);
+	if (curr_simos == SIMOS_CGC) {
+		assert (state->arch == UC_ARCH_X86);
+		assert (state->unicorn_mode == UC_MODE_32);
 
-			if (fd == 2) {
-				// we won't try to handle fd 2 prints here, they are uncommon.
-				return;
-			} else if (fd == 0 || fd == 1) {
-				uc_reg_read(uc, UC_X86_REG_ECX, &buf);
-				uc_reg_read(uc, UC_X86_REG_EDX, &count);
-				uc_reg_read(uc, UC_X86_REG_ESI, &tx_bytes);
-
-				// ensure that the memory we're sending is not tainted
-				// TODO: Can transmit also work with symbolic bytes?
-				void *dup_buf = malloc(count);
-				uint32_t tmp_tx;
-				if (uc_mem_read(uc, buf, dup_buf, count) != UC_ERR_OK)
-				{
-					//printf("... fault on buf\n");
-					free(dup_buf);
-					return;
-				}
-
-				if (tx_bytes != 0 && uc_mem_read(uc, tx_bytes, &tmp_tx, 4) != UC_ERR_OK)
-				{
-					//printf("... fault on tx\n");
-					free(dup_buf);
-					return;
-				}
-
-				if (state->find_tainted(buf, count) != -1)
-				{
-					//printf("... symbolic data\n");
-					free(dup_buf);
-					return;
-				}
-
-				state->step(state->transmit_bbl_addr, 0, false);
-				state->commit();
-				if (state->stopped)
-				{
-					//printf("... stopped after step()\n");
-					free(dup_buf);
-					return;
-				}
+		if (intno == 0x80) {
+			for (auto sr : state->symbolic_registers) {
+				// eax,ecx,edx,ebx,esi
+				if ((sr >= 8 && sr <= 23) || (sr >= 32 && sr <= 35)) return;
+			}
 
-				uc_err err = uc_mem_write(uc, tx_bytes, &count, 4);
-				if (tx_bytes != 0) state->handle_write(tx_bytes, 4, true);
-				if (state->stopped) {
-					return;
-				}
-				state->transmit_records.push_back({dup_buf, count});
-				int result = 0;
-				uc_reg_write(uc, UC_X86_REG_EAX, &result);
-				state->symbolic_registers.erase(8);
-				state->symbolic_registers.erase(9);
-				state->symbolic_registers.erase(10);
-				state->symbolic_registers.erase(11);
-				state->interrupt_handled = true;
-				state->syscall_count++;
-				return;
+			uint32_t sysno;
+			uc_reg_read(uc, UC_X86_REG_EAX, &sysno);
+			//printf("SYSCALL: %d\n", sysno);
+			if (sysno == state->cgc_transmit_sysno) {
+				state->perform_cgc_transmit();
+			}
+			else if (state->is_tracing_mode() && (sysno == state->cgc_receive_sysno)) {
```

Comment:
See above. If we don't want to handle receive in native land I think we should be explicit about that, perhaps by not providing the appropriate syscall number.

Reply:
For the receive syscall, it makes sense: I think we can get rid of this check. But, I think we will still need this check for some more enhancements that will be added such as handling symbolic memory addresses. During tracing, we have a concrete value to work with and we can use that but during full symbolic exploration, we will have to switch to VEX engine.